Issue 152 Accessing Idle API endpoint returns errors without context

chmouel opened the pull request:
Add a couple of tests along the way.

Some methods could not be tested (i.e: Info) since it needs some
refactoring. see fabric8-services/fabric8-jenkins-idler#169

kishansagathiya commented on internal/api/api.go:
```diff
@@ -135,8 +142,9 @@ func (api *idler) Info(w http.ResponseWriter, r *http.Request, ps httprouter.Par
 		if err != nil {
-			log.Error("Could not serialize users")
-			fmt.Fprint(w, "{'error': 'Could not serialize users'}")
+			log.Errorf("Could not serialize users: %s", err)
```
If we want to handle many errors in the same way or display a similar kind of formatted-error message,  probably we should create an error package with handling functions. WDYT?

chmouel replied:
I think it's a good idea, we can do this in another refactoring PR that can go all over the code for this, relevant [article](http://blog.questionable.services/article/http-handler-error-handling-revisited/) for this

kishansagathiya commented on internal/api/api.go:
```diff
@@ -155,5 +163,5 @@ func (api *idler) getURLAndToken(r *http.Request) (string, string, error) {
 		return openShiftAPIURL, bearerToken, nil
 	}
-	return "", "", nil
+	return "", "", fmt.Errorf("Unknown or invalid OpenShift API URL")
```
Shouldn't we return nil instead of empty strings? It's not critical, It doesn't break anything, you can skip it as well. I know throughout the project there are many functions like this. We can probably change them as we face them.
WDYT?

chmouel replied:
heu nil is not a str type compatible,

kishansagathiya replied:
but you can make it a pointer

chmouel replied:
I am not sure what's the point, I find "" much clearer (and that's why it's used like that all over go code),

kishansagathiya commented on internal/api/api_test.go:
```diff
@@ -0,0 +1,164 @@
+	functions := []ReqFuncType{mockidle.Idle, mockidle.UnIdle, mockidle.IsIdle}
+
+	for _, function := range functions {
+		r, _ := http.NewRequest("GET", "/", nil)
```
Shouldn't we use more descriptive names? `r` -> `request` ? Same goes for other variable names as well.

chmouel replied:
I don't really care personally, r is quite standard variable name for request in go world

kishansagathiya commented on internal/api/api_test.go:
```diff
@@ -0,0 +1,164 @@
+		w := &mockResponseWriter{}
+		function(w, r, nil)
+
+		assert.Equal(t, http.StatusBadRequest, w.writerStatus, fmt.Sprintf("Bad Error Code: %d", w.writerStatus))
```
shouldn't this be `require`?

kishansagathiya commented on internal/api/api_test.go:
```diff
@@ -0,0 +1,164 @@
+
+		ns := &JSError{}
+		_ = json.Unmarshal(w.buffer.Bytes(), &ns)
+		assert.Equal(t, "OpenShift API URL needs to be specified", ns.Error, fmt.Sprintf("Unexpected error output: %s", ns.Error))
```
shouldn't this be `require`?

chmouel replied:
why don't we want to continue to the other tests?

kishansagathiya replied:
If tests are sequential (If first fails, following ones are also likely to fail. Is that the case here?), require is a good idea.

chmouel replied:
I don't think it's that important, but if you feel that strongly I can change that,

chmouel commented:
Updated PR with some of your comments, moved all mocks to the `internal/testutils/mock` package, fixed another error where previously it wasn't reported properly, Before : 

![image](https://user-images.githubusercontent.com/98980/37669283-bfcf31cc-2c66-11e8-8c3c-42d636264ad8.png)

After : 

![image](https://user-images.githubusercontent.com/98980/37669278-bca398a8-2c66-11e8-8f9c-1daba47c2805.png)
